Add support of readWriteTokens

pjanik opened the pull request:
This PR adds complete support of readWriteTokens. They can be used both to create new resources or to obtain AWS keys for existing ones. Changes are in all three parts of this repo.

### functions
Most important changes here. There's a new accessRuleType - readWriteToken. Authorization isn't the same for all the routes anymore, so I had to move it from `app.use` to all the route handlers. It's defined in each route separately.

readWriteToken is created only when accessRuleType is `readWriteToken`. In such case, JWT is not required by the *create* route.

readWriteToken can be used to:
 - get AWS credentials
 - update token-service resource (name, description - not sure how useful it's, but seemed to make sense to be consistent)

readWriteToken **can't** be used to:
 - delete resource (my own assumption, but I think we don't want one owner of a shared resource to delete it for everyone else)

readWriteTokens have a predefined prefix, so the code can recognize them and differentiate from JWT tokens. Also, codes generated by token-service have second prefix, like "token-service-generated". When we migrate DocStore, we can add something like "doc-store" just in case this is ever useful (e.g. when we want to deprecate non-doc-store documents or migrate them to use JWT or something). It's not necessary, but it doesn't seem to cost anything.

Lots of the code around express app was using `Promise` class even if it was fully synchronous or could be. I guess it was done like that to keep style more consistent with the handler functions that were actually async. It works okay if functions are used only there. But now they are a bit more divided and used by few helpers in a different way. In such case, Promises are only annoying. I tried to use Promises where they could fit nicely the promises chain, but it leads to some mess sooner or later. So my rule of thumb was: don't use promises if the code is fully sync.

publicPath and publicUrl are now part of the token-service API responses and client logic has been updated. It was requested by @scytacki in the design doc. It isn't really necessary. But it makes things a bit clearer when you e.g. inspect response. 

### client

Changes:
 - `jwt` argument in constructor is optional
 - new method `getReadWriteToken` (does NOT talk to server) that extracts token from S3Resource object (e.g. returned by `getResource` or `createResource`). It's useful as token is hidden in `accessRules`.
 - `getCredentials` now also accepts optional parameter `readWriteToken`
 - `getPublicS3Path/Url` now use new props returned by API


### example-app

New helper `uploadFileAnonymously` shows how to create a new resource with readWriteToken, get this token, and how to get AWS credentials. Finally, it uploads a file content using the AWS SDK.

@scytacki, I think this PR generally follows your design described here:
https://docs.google.com/document/d/17QtyDzPoi_XOhGA4IAyO1OEz7zW0aaAqX5YBZKhA8Ug/edit?ts=5e946ccf#
Things that are different:
 - I didn't add S3SharedResource to keep things simple. It didn't seem necessary. We just use a new access rule type to S3Resource and things seem to fit nicely. But if you think S3SharedResource is useful for other reasons, let me know. I don't fully get these divisions, in particular, what is `iotOrganization` and how it's used. 
 - readWriteToken is stored in `accessRule` and it's also returned there by the API. Not in the top-level object if it was what you imagined. The client lib has a new method that extracts it easily. Generally token-service API always returns objects that it's working with (e.g. Firestore entries directly or AWS credentials), so I tried to keep that and don't extend them with some redundant information.

dougmartin commented on functions/src/index.ts:
```diff
@@ -46,163 +45,194 @@ const genericErrorHandler = (err: any, req: express.Request, res: express.Respon
-    .catch(error => res.error(400, error))
-})
+  try {
+    const claims = authUsingJWT(req);
```
You could also do these auth checks as a single express middleware function (like `checkEnv` above) that authed based on the path and then added the `claims` or `claimsOrReadWriteToken` to the `req` object.  I'm not sure which would be cleaner so I guess leave it as is.

pjanik replied:
I thought about that and even implemented something similar. We can also add middleware to particular handlers like:

`app.get('/api/v1/resources', authauthUsingReadWriteTokenOrJWT, (req, res) => {`
`app.get('/api/v1/resources', authUsingJWT, (req, res) => {`
`app.get('/api/v1/resources', authUsingReadWriteToken (req, res) => {`
etc.

The main problem is extending `request` object that gets messy long-term. Especially when we have properties that are optional and exclusive in fact: `claims` or `readWriteToken`. If we use this kind of middleware that sometimes assigns claims, sometimes read-write token, there's a problem with useful typings.

- We can mark these properties optional in the `Request` interface and we can explicitly check if they are present. Kinda safe, but then using middleware with error handling doesn't make sense, as we'll duplicate this logic (error handling). If auth rules changes, you need to update the code in two places. If you forget about handler, TS won't help, and users will see auth errors.
- We mark these properties optional and manually cast them in the handlers. We don't duplicate the error handling code. But typecasting (or `!` operator) is based only on the developer's knowledge. And if the implementation of auth middleware changes, types won't help (as we did casting or used `!`). I think it might be error-prone and doesn't seem clean.
- We mark them as always present in the `Request` interface. It's the same problem as above, we just don't have to cast things. But if auth changes for given method, TS won't detect an error.

So, explicit implementation in handlers seems the safest to me. Also, cleanest as we don't need additional `if`s, casting, or silent assumptions about things available in `req`.

scytacki replied:
Another thing to consider (and perhaps you already did), is to treat the claims more generically. So they aren't just JWT claims but they are authorization claims.  Then when there is a readWrite token, it is added a claims object with a claim like `{readWriteToken: [token]}`.  Then all the code can continue to just handle claims instead of needing to include claims or a string. So this would push all of the claim checking logic down into the BaseResourceObject

pjanik replied:
So there are two things here I think.

Yes, I was trying to implement `AuthClaims = JWTClaims | ReadWriteTokenClaims` 😉 
 ReadWriteTokenClaims was exactly `{readWriteToken: string}`. Surprisingly, it isn't much easier to work later with that as interfaces are not present in JS runtime ofc and I'd need to test for properties:
<img width="817" alt="Screenshot 2020-04-15 at 12 32 15" src="https://user-images.githubusercontent.com/767857/79328073-c7b83380-7f15-11ea-9132-106db12439f5.png">
<img width="545" alt="Screenshot 2020-04-15 at 13 12 44" src="https://user-images.githubusercontent.com/767857/79331040-d6551980-7f1a-11ea-8306-3b60f84648ec.png">

With string:
<img width="503" alt="Screenshot 2020-04-15 at 12 31 30" src="https://user-images.githubusercontent.com/767857/79328065-c4bd4300-7f15-11ea-9452-f2ac94707c37.png">

I just though the string version is a bit more bulletproof as we don't depend on property names etc. But I can change it, I don't think there's a big difference here.

> So this would push all of the claim checking logic down into the BaseResourceObject

This is in fact separate from string vs claims. We could have this generic auth middleware that generates AuthClaims or JWTClaims | string. But then it's not that clear which authentication method is required for given action. It'd be hidden in BaseResourceObject internals (or S3ResourceObject etc.). In the current implementation, we can see that more clearly I think, as handlers explicitly call authUsingJWT or authUsingJWTOrReadWriteToken. But again, that's matter of taste, I can change it if I'm missing something.

pjanik replied:
I ended up refactoring to AuthClaims approach, it's all pretty similar, but less wordy and it would be easier to add more methods in the future if we need to.

dougmartin commented on functions/src/index.ts:
```diff
@@ -46,163 +45,194 @@ const genericErrorHandler = (err: any, req: express.Request, res: express.Respon
+      .then(resource => res.success(resource.apiResult()))
+      .catch(error => res.error(400, error))
+  } catch (error) {
+
```
Did you mean this `catch` to be empty?

pjanik replied:
No. ;) Thanks.

pjanik commented on example-app/src/helpers.ts:
```diff
@@ -74,3 +74,41 @@ export const uploadFileUsingFirebaseJWT = async (fileContent: string, firebaseJw
+    description: "test file",
+    accessRuleType: "readWriteToken"
+  }) as S3Resource;
+  console.log("new resource:", resource);
```
I'm assuming example-app is meant to be used by developers. It was actually helpful for me to see what's going on. And I guess console.logs are fine, as that's handy, and I don't have to add an additional code that nicely displays JSON objects (chrome dev tools are doing that great).

pjanik commented on client/src/client.ts:
```diff
@@ -133,16 +138,19 @@ export class TokenServiceClient {
   }
 
-  private fetch(method: string, path: string, body?: any) {
+  private fetch(method: string, path: string, body?: any, readWriteToken?: string) {
```
Good point, I'll change that.

pjanik commented on example-app/src/helpers.ts:
```diff
@@ -74,3 +74,41 @@ export const uploadFileUsingFirebaseJWT = async (fileContent: string, firebaseJw
+
+export const uploadFileAnonymously = async (fileContent: string, tokenServiceEnv: "dev" | "staging") => {
+  const client = new TokenServiceClient({ env: tokenServiceEnv });
+  const filename = "test.txt";
```
It's just an example, so personally I don't care what the filename is (nor the file content). I can add it, it won't add too many lines of code.

scytacki commented on functions/src/index.ts:
```diff
@@ -46,163 +45,194 @@ const genericErrorHandler = (err: any, req: express.Request, res: express.Respon
 
 app.get('/api/v1/resources/:id', (req, res) => {
-  return BaseResourceObject.Find(db, req.env, req.params.id)
+  BaseResourceObject.Find(db, req.env, req.params.id)
```
This doesn't seem to be checking any auth. Doesn't this mean that someone who has the id of the resource would get the full object stored in firestore which would probably include the readWriteToken?

pjanik replied:
Ops, good catch. This was always open. So, I think there are two options:
1. We can lock it. But then the question is if we don't break some apps (glossary, vortex).
2. We can remove accessRules if user is not authorized.

pjanik replied:
So, I tried 2 in one of the new commits. You can take a look if it looks good to you. AccessRules will be stripped depending on auth claims. In the future, we can apply the same approach to other fields if necessary.

scytacki commented on functions/src/index.ts:
```diff
@@ -46,163 +45,194 @@ const genericErrorHandler = (err: any, req: express.Request, res: express.Respon
+    }
+  } else {
+    // Anonymous usage based on readWriteToken. NOT RECOMMENDED apart from testing. readWriteTokens are meant to
+    // be used mostly/only to support converted Document Store documents.
```
At this point I think we are going to stick with just using readWriteTokens for all shared documents. So the "not recommend" part of this should be removed.

scytacki commented on functions/src/index.ts:
```diff
@@ -46,163 +45,194 @@ const genericErrorHandler = (err: any, req: express.Request, res: express.Respon
-    .catch(error  => res.error(400, error))
-})
+  try {
+    const claims = authUsingJWT(req); // note that readWriteToken intentionally does NOT give access to delete method
```
Yes I think this is the right approach. I don't see a reason to allow readWriteTokens to allow deletion.

scytacki commented on functions/src/index.ts:
```diff
@@ -46,163 +45,194 @@ const genericErrorHandler = (err: any, req: express.Request, res: express.Respon
-    .catch(error => res.error(400, error))
-})
+  try {
+    const claimsOrReadWriteToken = authUsingJWTOrReadWriteToken(req);
```
I think it'd be safer to only allow JWT access to do a patch like this.
Otherwise, I think someone with a readWriteToken would be able to update the authRules which means they could mess up the readWrite token and lock themselves out.

pjanik replied:
I can change it. But the same can happen with JWT. A resource owner can remove access for himself. ;)

pjanik replied:
OK, that's restricted now.

scytacki commented on functions/src/resource.test.ts:
```diff
@@ -189,6 +212,20 @@ describe("Resource", () => {
     });
 
+    describe("#isReadWriteTokenValid", () => {
+      it("should fail if tokes don't use ReadWriteTokenPrefix", () => {
```
```suggestion
      it("should fail if tokens don't use ReadWriteTokenPrefix", () => {
```

scytacki commented on functions/src/resource.ts:
```diff
@@ -131,81 +149,107 @@ export class BaseResourceObject implements BaseResource {
-                reject(`Unknown resource type: ${type}`);
+      return BaseResourceObject.GetResourceSettings(db, env, type, tool)
+        .then((settings) => {
+          let newResource: FireStoreResource;
```
I think we should allow tools to specify if they allow readWriteToken access rules. Otherwise someone could create a readWriteToken access to the glossary, vortex data, or an IoT System.

Perhaps this is a reason to add a new resource type of s3FolderShared (or something like that). Then this could be done by limiting the tool to a resource type. Which could then be used to improve the conditional on line 216: `if (!(tool === "dataFlow")) {`

pjanik replied:
There's a new "allowedAccessRuleTypes" array in settings now. 

Line 216 has been removed as settings define both tool and resource type as you noticed. So the code will fail higher up.

scytacki commented on functions/src/resource.ts:
```diff
@@ -131,81 +149,107 @@ export class BaseResourceObject implements BaseResource {
 
             const resource = BaseResourceObject.FromDocumentSnapshot(docSnapshot);
-            if (resource.isOwner(claims)) {
+            if (resource.canUpdate(claimsOrReadWriteToken)) {
```
This is where it seems unsafe to allow a readWriteToken request to update the resource. On line 257 they could change the access rules and lock themselves out.

scytacki commented on functions/src/resource.ts:
```diff
@@ -286,16 +330,48 @@ export class S3ResourceObject extends BaseResourceObject {
 
-  apiResult(): S3ResourceObject {
-    const result = super.apiResult() as S3ResourceObject;
+  get publicPath() {
```
Not necessary in this PR but it seems it would be helpful to put the S3ResourceObject class in its own file. I think it would make looking at the code easier.

pjanik replied:
Yeap, I'll do it later not to add more diff to this PR.

scytacki commented on functions/src/resource.ts:
```diff
@@ -36,7 +39,12 @@ export class BaseResourceObject implements BaseResource {
     return {id, name, description, type, tool, accessRules};
```
This seems to be the place where the accessRules for the resource would be returned in a get request. So it seems the get request needs to be authenticated otherwise it will expose the readWrite token.  I think this might conflict with some other usage of the get request, so perhaps it requires more discussion.

pjanik commented:
@scytacki, @dougmartin, I think it's ready for the second review now.
Commit names should show what I have changed. I think I addressed most of the comments (or replied to them). I tried to simplify authentication, especially the "optional" path.

pjanik commented on client/src/client.ts:
```diff
@@ -102,27 +103,26 @@ export class TokenServiceClient {
   }
 
+  getReadWriteToken(resource: S3Resource): string | undefined {
+    return getReadWriteToken(resource);
```
Right. It's just delegating to the helper that is also used by the server-side code. I renamed it to `getRWTokenFromAccessRules`.

pjanik commented:
Thanks, I pushed the suggested changes.
